src/validation.cpp

```diff
@@ -2512,45 +2512,20 @@ static int64_t nTimeFlush = 0;
 static int64_t nTimeChainState = 0;
 static int64_t nTimePostConnect = 0;
 
+//! A single block and block index pointer.
 struct PerBlockConnectTrace {
     CBlockIndex* pindex = nullptr;
     std::shared_ptr<const CBlock> pblock;
-    PerBlockConnectTrace() {}
+    PerBlockConnectTrace(CBlockIndex* index, std::shared_ptr<const CBlock> block) : pindex(index), pblock(block) {}
```

Comment:
```suggestion
    PerBlockConnectTrace(CBlockIndex* index, const std::shared_ptr<const CBlock>& block) : pindex(index), pblock(block) {}
```

style nit to avoid shared_ptr increment/decrement

Reply:
Thanks for this comment! It sent me down a bit of a rabbit hole about how to pass smart pointers. Here's what I've learned:

I think your suggestion to change the ctor to take a const lvalue reference makes a copy and therefore increases the ref count (the `const` is the callee saying to the caller "I won't steal this reference", and so if the callee wants to keep ownership, it needs to make a copy). We can avoid that copy in three ways:

- pass by non-const lvalue reference (`std::shared_ptr<const CBlock>&`) which is the callee saying "I may or may not steal this reference" and call the constructor with an lvalue (ie `thisBlock` below).
- pass by value (`std::shared_ptr<const CBlock>`), and have the caller cast to rvalue reference in the call (ie `std::move(thisBlock)` below). The `std::move` is the caller saying "I no longer want ownership of this reference. In this case, the caller can still keep shared ownership by passing an lvalue (`thisBlock`), which would increase the ref count.
- pass by rvalue (`std::shared_ptr<const CBlock>&&`) which is the callee saying "I will steal ownership". The caller has to call using an rvalue reference (`std::move(thisBlock)`). Calling by lvalue would be invalid.

I think it's best to keep this as pass-by-value and have the caller pass it with an rvalue reference.

Also, the initializer list needs to use a std::move to prevent the ref increment/decrement, which I'll change.

Some references:
- summary of ways to pass smart pointers as arguments (it's talking about unique_ptr, but the same concepts are true for shared_ptrs): https://stackoverflow.com/questions/8114276/how-do-i-pass-a-unique-ptr-argument-to-a-constructor-or-a-function
- CPP Core Guidelines on passing shared pointers: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-sharedptrparam-owner
- Herb Sutter on smart pointer parameters: https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/
- Scott Meyers on passing move types by value/rvalue reference: http://scottmeyers.blogspot.com/2014/07/should-move-only-types-ever-be-passed.html
- Demonstration code of the different ways to pass a shared_ptr and the effect on ref counts: https://gist.github.com/jnewbery/10ee24a5ce783932accddbe4347f110e